staff/demi/spotify-app-react/script/main.jsx

```diff
@@ -0,0 +1,358 @@
+
+        this.state = {
+            artistsSearched: [],
+            showArtists: true,
+            showAlbums: false,
+            albumsSearched: [],
+            showSongs: false,
+            songsSearched: [],
+            songSelected:"",
+            showSong:false,
+        }
+    }
+
+
+    addArtist = (artist) => {
+
+        spotifyApi.searchArtists(artist).then(artists => {
+
+            artists = artists.map((artist) => {
+
+                if (artist.images.length === 0) {
+                    artist.images[0] = {
+                        url: "http://playercdn.listenlive.co/templates/StandardPlayerV4/webroot/img/default-cover-art.png"
+                    }
+                }
+                return artist
+
+            })
+            this.setState({ artistsSearched: artists })
+        })
+    }
+
+
+    retrieveAlbums = (e) => {
+
+        this.setState({ showArtists: false })
+        this.setState({ showAlbums: true })
+
+        var id = e.target.getAttribute('data')
+
+        spotifyApi.retrieveAlbums(id).then(albums => {
+
+            albums = albums.map((album) => {
+                return album
+            })
+
+            this.setState({ albumsSearched: albums })
+
+        })
+    }
+
+
+    retrieveTracks = (e) => {
+
+        this.setState({ showAlbums: false })
+        this.setState({ showSongs: true })
+
+        var href = e.target.getAttribute('data-url') + "/tracks"
+
+        spotifyApi.retrieveTracks(href).then(songs => {
+
+            songs = songs.map((song) => {
+                return song
+            })
+
+            this.setState({ songsSearched: songs })
+        })
+
+
+    }
+
+
+    retrieveTrack = (e) => {
+
+        this.setState({ showSong: true })
+
+        var song = e.target.getAttribute('data')
+
+        spotifyApi.retrieveTrack(song).then(song =>{
+
+            this.setState({ songSelected: song })
+
+        })
+
+    }
+
+
+
+
+    render() {
+        return <div>
+
+            <SpotyInput
+                onAddArtist={this.addArtist}
+            />
+
+            {this.state.showArtists ? <SpotyCovers artistsSearched={this.state.artistsSearched} retrieveAlbums={this.retrieveAlbums} /> : null}
+
+            {this.state.showAlbums ? <SpotyAlbums retrieveAlbums={this.retrieveAlbums} albumsSearched={this.state.albumsSearched} retrieveTracks={this.retrieveTracks} /> : null}
+
+            <SpotySongs
+                retrieveTracks={this.retrieveTracks}
+                songsSearched={this.state.songsSearched}
+
+                retrieveTrack={this.retrieveTrack}
+            />
+
+            {this.state.showSong ? <SpotySong retrieveTrack={this.retrieveTrack} songSelected={this.state.songSelected}/>: null}
+
+
+        </div>
+
+    }
+}
+
+
+
+class SpotyInput extends React.Component {
+    constructor() {
+        super()
+
+        this.state = {
+            query: ''
+        }
+
+    }
+
+    keepInput = e => this.setState({ query: e.target.value })
```

Comment:
MOVE event (e) treating to the input element instead of locating it here. so keepInput just only receives the input value, but not the event. that is:

keepInput = query => this.setState({ query})

<input onChange={e => this.keepInput(e.target.value)} ... />